File: code/richediting/src/main/java/org/fbme/ide/richediting/viewmodel/NetworkConnectionView.java
package org.fbme.ide.richediting.viewmodel;

import org.fbme.ide.iec61499.adapter.common.PlatformElement;
import org.fbme.lib.iec61499.fbnetwork.ConnectionPath;
import org.fbme.lib.iec61499.fbnetwork.EntryKind;
import org.fbme.lib.iec61499.fbnetwork.FBNetworkConnection;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.mps.openapi.model.SNode;

import java.util.Objects;

public class NetworkConnectionView {

    @NotNull
    private final Object myConnection;
    private ConnectionPath myConnectionPath;
    private final boolean myEditable;
    private final SNode myAssociatedNode;
    private final EntryKind myKind;

    public NetworkConnectionView(@NotNull Object connection, boolean editable) {
        myConnection = connection;
        myConnectionPath = connection instanceof FBNetworkConnection ? ((FBNetworkConnection) connection).getPath() : new ConnectionPath();
        myKind = connection instanceof FBNetworkConnection ? ((FBNetworkConnection) connection).getKind() : EntryKind.DATA;
        myEditable = editable;
        myAssociatedNode = ((PlatformElement) connection).getNode();
    }

    public SNode getAssociatedNode() {
        return myAssociatedNode;
    }

    public EntryKind getKind() {
        return myKind;
    }

    public void shrink() {
        myConnectionPath = new ConnectionPath();
    }

    public void setPath(ConnectionPath path) {
        FBNetworkConnection connection = getConnection();
        if (myEditable && connection != null) {
            connection.setPath(path);
        }
    }

    @Nullable
    public FBNetworkConnection getConnection() {
        return myConnection instanceof FBNetworkConnection ? (FBNetworkConnection) myConnection : null;
    }

    public boolean isEditable() {
        return myEditable;
    }

    @NotNull
    public ConnectionPath getConnectionPath() {
        return myConnectionPath;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof NetworkConnectionView)) {
            return false;
        }

        NetworkConnectionView that = (NetworkConnectionView) o;
        return Objects.equals(myConnection, that.myConnection);
    }

    @Override
    public int hashCode() {
        return myConnection.hashCode();
    }
}

File: code/richediting/src/main/kotlin/org/fbme/ide/richediting/viewmodel/NetworkConnectionView.kt
package org.fbme.ide.richediting.viewmodel

import org.fbme.ide.iec61499.repository.PlatformElement
import org.fbme.lib.iec61499.declarations.ParameterAssignment
import org.fbme.lib.iec61499.fbnetwork.ConnectionPath
import org.fbme.lib.iec61499.fbnetwork.EntryKind
import org.fbme.lib.iec61499.fbnetwork.FBNetworkConnection
import org.jetbrains.mps.openapi.model.SNode

class NetworkConnectionView {
    private val myConnection: Any
    var connectionPath: ConnectionPath
        private set
    val isEditable: Boolean
    val associatedNode: SNode
    val kind: EntryKind

    constructor(connection: FBNetworkConnection, editable: Boolean) {
        myConnection = connection
        connectionPath = connection.path!!
        kind = connection.kind
        isEditable = editable
        associatedNode = (connection as PlatformElement).node
    }

    constructor(parameter: ParameterAssignment, editable: Boolean) {
        myConnection = parameter
        connectionPath = ConnectionPath()
        kind = EntryKind.DATA
        isEditable = editable
        associatedNode = (parameter as PlatformElement).node
    }

    fun shrink() {
        connectionPath = ConnectionPath()
    }

    fun setPath(path: ConnectionPath?) {
        val connection = connection
        if (isEditable && connection != null) {
            connection.path = path
        }
    }

    val connection: FBNetworkConnection?
        get() = if (myConnection is FBNetworkConnection) myConnection else null

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other !is NetworkConnectionView) {
            return false
        }
        return myConnection == other.myConnection
    }

    override fun hashCode(): Int {
        return myConnection.hashCode()
    }
}
